File: code/backend/java/adapter-auth/src/main/java/games/paths/adapters/auth/persistence/UserAccessAdapter.java
package games.paths.adapters.auth.persistence;

import games.paths.adapters.auth.entity.UserEntity;
import games.paths.adapters.auth.repository.UserRepository;
import games.paths.core.port.match.UserAccessPort;

import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

/**
 * UserAccessAdapter - Adapter that exposes a read-only view of the
 * {@code users} table to the match domain via {@link UserAccessPort}.
 * Step 19 — used to validate the creator on POST /matches and to
 * resolve the user uuid on GET /match/&#123;uuid&#125;/info.
 */
@Repository
@Transactional(readOnly = true)
public class UserAccessAdapter implements UserAccessPort {

    private final UserRepository userRepository;

    public UserAccessAdapter(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public Optional<UserView> findByUuid(String uuid) {
        if (uuid == null || uuid.isBlank()) {
            return Optional.empty();
        }
        return userRepository.findByUuid(uuid).map(this::toView);
    }

    private UserView toView(UserEntity entity) {
        return new UserView(
                entity.getId(),
                entity.getUuid(),
                entity.getUsername(),
                entity.getRole(),
                entity.getState());
    }
}

File: code/backend/python/app/adapters/persistence/match/user_access_adapter.py
"""Step 19 — :class:`UserAccessPort` adapter against the SQLAlchemy users table."""
from typing import Any, Dict, Optional

from app.adapters.persistence.auth.models import User
from app.core.ports.match.match_ports import UserAccessPort


class UserAccessAdapter(UserAccessPort):
    def __init__(self, session_factory):
        self.session_factory = session_factory

    def find_by_uuid(self, user_uuid: str) -> Optional[Dict[str, Any]]:
        if not user_uuid:
            return None
        with self.session_factory() as session:
            entity = session.query(User).filter(User.uuid == user_uuid).first()
            if entity is None:
                return None
            return {
                "id": entity.id,
                "uuid": entity.uuid,
                "username": entity.username,
                "role": entity.role,
                "state": entity.state,
            }
